### PR TITLE
Niagara+ update!

### PR DESCRIPTION
**Pull request by orsenkucher**
## Status
**READY**

## Breaking Changes
NO

## Description
From now using `hive` package as core of storage, backwards compatible.

Basically I've added hive support based on `water` package, removed all multi-file stuff, but left reworked single file storage for compatibility reasons.
One thing to note, any serialization was moved from `hydrated_bloc.dart`

## Todos
- [x] Tests (no new added, old updated)
- [x] Documentation

## To run break test
```bash
git checkout break-test-v3
cd breaktest/
flutter run
git checkout break-test-v4.1
flutter run
```

### CODE REVIEW

File: lib/src/hydrated_bloc_delegate.dart [hunk 1]
```diff
@@ -14,17 +15,32 @@ class HydratedBlocDelegate extends BlocDelegate {
...
+  /// import 'package:crypto/crypto.dart';
+  /// import 'package:hive/hive.dart';
+  ///
+  /// const password = 'hydration';
+  /// final byteskey = sha256.convert(utf8.encode(pass)).bytes;
+  /// return HiveAesCipher(byteskey);
+  /// ```
   static Future<HydratedBlocDelegate> build({
     Directory storageDirectory,
+    HiveCipher encryptionCipher,
```
Comment (felangel) [inline lib/src/hydrated_bloc_delegate.dart]: I think it would be nice to abstract this so the public api doesn't expose direct dependencies on hive in case we pivot to another storage solution in the future. Thoughts?

File: lib/src/hydrated_bloc_storage.dart [hunk 2]
```diff
@@ -20,69 +22,171 @@ abstract class HydratedStorage {
...
+  /// import 'package:crypto/crypto.dart';
+  /// import 'package:hive/hive.dart';
+  ///
+  /// const password = 'hydration';
+  /// final byteskey = sha256.convert(utf8.encode(pass)).bytes;
+  /// return HiveAesCipher(byteskey);
+  /// ```
   static Future<HydratedBlocStorage> getInstance({
     Directory storageDirectory,
+    HiveCipher encryptionCipher,
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: same comment here about abstracting the `Hive` implementation details.

File: lib/src/hydrated_bloc_storage.dart [hunk 3]
```diff
@@ -20,69 +22,171 @@ abstract class HydratedStorage {
...
-          storage =
-              json.decode(await file.readAsString()) as Map<String, dynamic>;
-        } on dynamic catch (_) {
-          await file.delete();
+      if (!kIsWeb) {
+        Hive.init(directory.path);
+      }
+
+      final box = await Hive.openBox(
+        'niagara+',
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]:
```suggestion
        'hydrated_box',
```
thoughts?

Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: 😆

File: lib/src/hydrated_bloc_storage.dart [hunk 4]
```diff
@@ -20,69 +22,171 @@ abstract class HydratedStorage {
...
+      _instance = null;
+      return _lock.synchronized(_box.deleteFromDisk);
+    } else {
+      return null;
+    }
   }
 }
+
+/// `CellSinglet` is storage of one file.
+class CellSinglet {
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: do you still think there's value in maintaining a notion of a CellSinglet/Cell etc if we're only supporting single file for backward compat? I kind of feel it would be simpler to remove the abstract because it's a bit hard to follow in my opinion...could just be because I'm tired though 😛 

Let me know what you think 👍

File: test/hydrated_storage_test.dart [hunk 5]
```diff
@@ -163,22 +163,24 @@ void main() {
...
               i,
               (i) => Iterable.generate(i, (j) => 'Point($i,$j);').toList(),
             ).toList();
+
             hydratedStorage.write(token, record); // no await here
 
             hydratedStorage = await HydratedBlocStorage.getInstance(
               storageDirectory: directory,
-            ); // basically refreshes cache
+            ); // basically does nothing now
```
Comment (felangel) [inline test/hydrated_storage_test.dart:178]: if it does nothing can we remove it haha?

**Review by felangel — CHANGES_REQUESTED** [review]
Looks great overall! Just have a few questions I want to talk through. Let me know what you think 👍

File: lib/src/hydrated_bloc_storage.dart (hunk 4 above)
Comment (orsenkucher) [reply lib/src/hydrated_bloc_storage.dart]: Yep, I was like stripping code I already had. Good idea to boil it down even more👍

File: test/hydrated_storage_test.dart (hunk 5 above)
Comment (orsenkucher) [reply test/hydrated_storage_test.dart:178]: idk, it can be useful if we introduce a way to create custom boxes: some for encrypted bloc's, some for overall separation of concerns

Comment (felangel) [reply test/hydrated_storage_test.dart:178]: alright fair enough 👍

File: lib/src/hydrated_bloc_delegate.dart [hunk 6]
```diff
@@ -14,17 +14,32 @@ class HydratedBlocDelegate extends BlocDelegate {
...
-  /// default `HydratedBlocStorage`.
-  /// A custom `storageDirectory` can optionally be provided.
+  /// default [HydratedBlocStorage].
+  /// A custom [storageDirectory] can optionally be provided.
   ///
   /// This is the recommended way to use a `HydratedBlocDelegate`.
   /// If you want to customize `HydratedBlocDelegate`
   /// you can extend `HydratedBlocDelegate` and perform the necessary overrides.
+  ///
+  /// [encryptionCipher] is Hive's `HiveCipher`,
```
Comment (felangel) [inline lib/src/hydrated_bloc_delegate.dart]: can we update the docs to reflect the new `HydratedCipher`?

File: lib/src/hydrated_bloc_storage.dart [hunk 7]
```diff
@@ -20,69 +24,107 @@ abstract class HydratedStorage {
...
-  final Map<String, dynamic> _storage;
-  final File _file;
-
+class HydratedBlocStorage extends HydratedStorage {
   /// Returns an instance of `HydratedBlocStorage`.
-  /// `storageDirectory` can optionally be provided.
+  /// [storageDirectory] can optionally be provided.
   /// By default, `getTemporaryDirectory` is used.
+  ///
+  /// [encryptionCipher] is Hive's `HiveCipher`,
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: can we update the docs to reflect the new `HydratedCipher`?

Comment (orsenkucher) [reply lib/src/hydrated_bloc_storage.dart]: yep, noticed that too

**Review by felangel — APPROVED** [review]
Looks great to me! Thanks again for the awesome contribution! 💯 🚀
